### PR TITLE
Dawn engine: refactor

### PR DESCRIPTION
**Pull request by dneto0**
- Eliminate lots of state from the Dawn engine
  - The Dawn Queue
    - Create and destroy within a single Amber command execution instead.
  - The Dawn CommandBufferEncoder:
    - Create and destroy within a single Amber command execution instead.
- Map an Amber Pipeline object to a Dawn-engine Pipeline object
- Use the framebuffer width and height from the Amber pipeline object,
  instead of hardcoding it.
- The framebuffer buffer is only mapped during the execution of an Amber
  command, then immediately unmapped afterward.

### CODE REVIEW

**Comment by dneto0** [comment]
Hopefully fixed the clang-format check, which was formatting differences and also an include-what-you-use violation

File: src/dawn/engine_dawn.cc [hunk 1]
```diff
@@ -259,86 +260,103 @@ Result EngineDawn::SetShader(ShaderType type,
...
+  // TODO(dneto): Likely, we can create the render objects during
+  // CreatePipeline.
+  Result result = CreateRenderObjectsIfNeeded(render_pipeline);
   if (!result.IsSuccess())
     return result;
 
-  // TODO(sarahM0): Refactor this code later.  This backend is in transition
-  // between the old design and the new design of how Amber operates.
+  // TODO(sarahM0): Adapt this to the case where there might be many colour
+  // attachments. Also whne there is a depth/stencil attachment.
```
Comment (dj2) [inline src/dawn/engine_dawn.cc]: nit: when

File: src/dawn/engine_dawn.cc [hunk 2]
```diff
@@ -417,64 +437,52 @@ Result EngineDawn::DoBuffer(const BufferCommand*) {
...
-        "EngineDawn: Can't create command buffer builder: device is not "
-        "initialized");
-  }
-  command_buffer_builder_ = device_->CreateCommandEncoder();
-  if (command_buffer_builder_)
-    return {};
-  return Result("EngineDawn: Can't create command buffer builder");
+Result EngineDawn::CreateRenderObjectsIfNeeded(
+    RenderPipelineInfo* render_pipeline) {
+  return CreateFramebufferIfNeeded(render_pipeline);
```
Comment (dj2) [inline src/dawn/engine_dawn.cc]: Do we expect to do more here? Or else, why not just call CreateFramebufferIfNeeded directly to make it clearer whta's happening?

File: src/dawn/engine_dawn.cc [hunk 3]
```diff
@@ -417,64 +437,52 @@ Result EngineDawn::DoBuffer(const BufferCommand*) {
...
 
-Result EngineDawn::CreateFramebufferIfNeeded() {
-  Result result;
+  // First make the Dawn color attachment textures that the render pipeline will
+  // write into.
   {
     ::dawn::Texture fb_texture;
-    result = MakeFramebufferTexture(*device_, kFramebufferFormat, &fb_texture);
+
+    result = MakeFramebufferTexture(*device_, kFramebufferFormat, width, height,
```
Comment (dj2) [inline src/dawn/engine_dawn.cc]: The framebuffer format should come from the pipeline, the format is specified in the script.

File: src/dawn/engine_dawn.h [hunk 4]
```diff
@@ -61,31 +61,32 @@ class EngineDawn : public Engine {
   Result DoBuffer(const BufferCommand* cmd) override;
 
  private:
-  // Creates a command buffer builder if it doesn't already exist.
-  Result CreateCommandBufferBuilderIfNeeded();
-  // Destroys the current command buffer builder.
-  void DestroyCommandBufferBuilder();
+  // Returns the Dawn-specific render pipepline for the given command,
```
Comment (dj2) [inline src/dawn/engine_dawn.h]: nit: pipeline

File: src/dawn/pipeline_info.h [hunk 5]
```diff
@@ -62,12 +66,20 @@ struct RenderPipelineInfo {
 
 struct ComputePipelineInfo {
   ComputePipelineInfo() {}
-  explicit ComputePipelineInfo(::dawn::ShaderModule comp)
-      : compute_shader(comp) {}
+  explicit ComputePipelineInfo(::amber::Pipeline* the_pipeline,
```
Comment (dj2) [inline src/dawn/pipeline_info.h]: explicit is only needed if there is only one parameter

File: src/dawn/engine_dawn.cc (hunk 3 above)
Comment (dneto0) [reply src/dawn/engine_dawn.cc]: done in most recent commit.  This is the biggest change to what you've already seen.

File: src/dawn/pipeline_info.h (hunk 5 above)
Comment (dneto0) [reply src/dawn/pipeline_info.h]: removed.

**Comment by dneto0** [comment]
Hey @dj2 and @sarahM0 this is ready for re-review.   Getting the framebuffer format from the Amber pipeline object was kind of a big change.

This is also rebased against latest master.

**Pull request merged**